core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,196 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InterruptedIOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    static <T> StreamMessageInputStream<T> of(
+            StreamMessage<T> source, Function<? super T, ? extends HttpData> httpDataConverter) {
+        return new StreamMessageInputStream<>(source, httpDataConverter);
+    }
+
+    private final StreamMessage<T> source;
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+    private final AtomicBoolean subscribed = new AtomicBoolean();
+    @Nullable
+    private volatile InputStream inputStream;
+    private volatile boolean closed;
+
+    private StreamMessageInputStream(StreamMessage<T> source,
+                                     Function<? super T, ? extends HttpData> httpDataConverter) {
+        requireNonNull(source, "source");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        this.source = source;
+        subscriber = new StreamMessageInputStreamSubscriber<>(httpDataConverter);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return read(in -> {
+            try {
+                return in.read();
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        return read(in -> {
+            try {
+                return in.read(b, off, len);
+            } catch (IOException e) {
+                return Exceptions.throwUnsafely(e);
+            }
+        });
+    }
+
+    private int read(Function<InputStream, Integer> function) throws IOException {
+        checkClosed();
+        if (subscribed.compareAndSet(false, true)) {
+            source.subscribe(subscriber);
+        }
+        if (inputStream == null || inputStream.available() == 0) {
+            try {
+                inputStream = subscriber.nextStream();
+            } catch (InterruptedException e) {
+                final InterruptedIOException ioe = new InterruptedIOException();
+                ioe.initCause(e);
+                throw ioe;
+            }
+        }
+        return function.apply(inputStream);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (closed) {
+            return;
+        }
+        closed = true;
+        subscriber.cancel();
```

Comment:
> Should we close the referenced inputStream if it is not null?

✅ updated

> How about using source.abort() to avoid an additional delay

✅ updated